lib/propolis/src/hw/virtio/softnpu.rs

```diff
@@ -0,0 +1,1061 @@
+///
+/// When a SoftNpu device is instantiated there is no P4 program that runs by
+/// default. A program must be loaded onto the emulated ASIC just like a real
+/// ASIC. This is accomplished through the P9 file system device exposed by
+/// SoftNpu. This P9 implementation exports a specific version string 9P2000.P4
+/// and only implements file writes to allow a consumer to upload a P4 program.
+///
+/// SoftNpu takes pre-compiled P4 programs in the form of shared libraries.
+/// These shared libraries must export a [pipeline constructor](
+/// https://oxidecomputer.github.io/p4/p4rs/trait.Pipeline.html) under the
+/// symbol `_main_pipeline_create`. Programs compiled with the `x4c` compiler
+/// export this symbol automatically.
+///
+/// Once a pre-compiled P4 program is loaded, the Pipeline object from that
+/// program is used to process packets. SoftNpu uses the illumos dlpi interface
+/// to send and receive raw Ethernet frames from the data link devices it has
+/// been configured with. Each frame received is processed with the loaded
+/// pipeline. If the pipeline invocation returns an egress port, then the egress
+/// packet returned by the pipeline will be sent out that port. If no egress
+/// port is returned, the packet is dropped.
+///
+/// In addition to forwarding packets between ports, SoftNpu also supports
+/// forwarding packets to and from the guest. This is accomplished through a
+/// special `pci_port` device. This is a viona device that shows up in the guest
+/// as a virtio network device. When a pipeline invocation returns an egress
+/// port of `0`, packets are sent to this port.
+///
+/// Most P4 programs require a corresponding control plane program to manage
+/// table state. For example, a program to add routing entries onto the ASIC. P4
+/// programs themselves only handle packets, they are not capable of managing
+/// table state. SoftNpu provides a uart-based management interface so that
+/// programs running in the guest can modify the tables of the P4 program loaded
+/// onto the ASIC. This is uart plumbed into the guest as `tty03`. What tables
+/// exist and how they can be modified is up to the particular program that is
+/// loaded. SoftNpu just provides a generic interface for table management and a
+/// few other ASIC housekeeping items like determining the number of ports.
+pub struct SoftNpu {
+    /// Data links SoftNpu will hook into.
+    pub data_links: Vec<String>,
+
+    /// dlpi handles for data links.
+    pub data_handles: Vec<dlpi::DlpiHandle>,
+
+    /// The PCI port.
+    pub pci_port: Arc<PciVirtioSoftNpuPort>,
+
+    /// Virtio state for CPU port.
+    virtio: Arc<PortVirtioState>,
+
+    /// UART for management from guest
+    uart: Arc<LpcUart>,
+
+    /// P9 file system endpoint for pre-compiled program transfer
+    pub p9fs: Arc<PciVirtio9pfs<SoftNpuP9Handler>>,
+
+    //TODO should be able to do this as a RwLock
+    pipeline: Arc<Mutex<Option<(Library, Box<dyn Pipeline>)>>>,
```

Comment:
Given how `data_handles`, the (CPU port) virtio state, and pipeline are all used by the worker threads, does it make sense to keep extra copies here (where they are passed to said workers during spawn), or perhaps just hand off the `Arc<PciVirtioSoftNpuPort>`, into which those workers can reach for the bits they need?

Reply:
This is all tucked away in the `PciVirtioSoftnpuPort` now, copies have been removed from `SoftNpu`.